### DISCUSSION TITLE
How can I use an UTF-8 font in cosmos?

### DISCUSSION THREAD

**Discussion by MasterYuan418** [CosmosOS/Cosmos#2384] [Q&A]
As the title says, I'm trying to load a font that includes Chinese.Can I just load it directly?

**Comment by terminal-cs** [comment]
Normal fonts (ttf, otf, etc) are not supported normally in cosmos, however there are many projects in cosmos which allow you to load those kinds of files, from most creators you can use it in your project with no issue, most won't mind it. Here's a good starting point: https://github.com/Relfos/LunarFonts

**Reply by MasterYuan418** [reply]
No, it report `native code encountered, plug required.`. Anyway, thanks a lot.

**Reply by terminal-cs** [reply]
> No, it report `native code encountered, plug required.`. Anyway, thanks a lot.

It shouldn't do this, you must have done something incorrectly because the code for it is quite simple. What have you changed in it and how did you try using it?

**Reply by MasterYuan418** [reply]
I used it with:
```csharp
            var fbytes = Files.Font; // Font is byte[] array with .TTF file
            var font = new Font(fbytes);
            var scale = font.ScaleInPixels(64);
            font.RenderGlyph('嗨', scale);
```

**Reply by MasterYuan418** [reply]
Also, that's my build return:
```
1>Native code encountered, plug required.
1>  DO NOT REPORT THIS AS A BUG.
1>  Please see http://www.gocosmos.org/docs/plugs/missing/
1>  Need plug for: System.Void  System.RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter(System.Runtime.CompilerServices.QCallTypeHandle, System.IntPtr*, System.Int32, System.Runtime.CompilerServices.ObjectHandleOnStack)(Plug Signature: System_Void__System_RuntimeTypeHandle_CreateInstanceForAnotherGenericParameter_System_Runtime_CompilerServices_QCallTypeHandle__System_IntPtr#__System_Int32__System_Runtime_CompilerServices_ObjectHandleOnStack_ ).
1>  Static: True
1>  Assembly: System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
```

**Reply by retrievaaaa** [reply]
I'd like to mention that

1. LunarFonts requires some changes
2. LunarFonts is incredibly broken. Many fonts don't render corrrectly.

I modified LunarFonts to be compatible, and made it public as CosmosTTF, but then decide to take it down as its completely broken with common fonts like Arial.

Your best bet is to port the original C version LunarFonts is based on (LunarFonts is about 8 years behind on fixes & features that have been added to the C version)

**Reply by retrievaaaa** [reply]
https://github.com/GoldenretriverYT/CosmosTTF

There you go, I made it public again. But as I said, it doesn't work with every TTF font

**Reply by AsertCreator** [reply]
cosmos does not support neither ttf/otf nor unicode
